core/src/test/java/hudson/model/AbstractItemTest.java

```diff
@@ -91,4 +94,41 @@ public void testSetDisplayNameOrNull() throws Exception {
         assertEquals(displayName, i.getDisplayNameOrNull());
         assertEquals(displayName, i.getDisplayName());
     }
+
+    private class NameNotEditableItem extends AbstractItem {
+
+        protected NameNotEditableItem(ItemGroup parent, String name){
+            super(parent, name);
+        }
+
+        @Override
+        public Collection<? extends Job> getAllJobs() {
+            return null;
+        }
+
+        @Override
+        public boolean isNameEditable() {
+            return false; //so far it's the default value, but it's good to be explicit for test.
+        }
+    }
+
+    @Test
+    @Issue("JENKINS-58571")
+    public void renameMethodShouldThrowExceptionWhenNotIsNameEditable() {
+
+        //GIVEN
+        NameNotEditableItem item = new NameNotEditableItem(null,"NameNotEditableItem");
```

Comment:
This is a pretty low-level test. That makes it fast, but on the other hand it is less clear that it really prevents the bug. A stronger test would actually make a real item (`createProject`) and try to use the `confirmRename` REST endpoint using `WebClient`. That however would require having an item type in the test classpath which is not a `Job` and which does not support renames, which would be trickier.